### DISCUSSION TITLE
[WIRESHARK] help decoding websocket exchanges with EMQX 5.1

### DISCUSSION THREAD

**Discussion by ldamet** [emqx/emqx#11156] [Q&A] [answered]
Hi everybody,

I am investigating some issues regarding some mqtt development. I have created a test environment in which I have 2 separate MQTT brokers:

* one docker EMQX 5.1 exposing port 8083
* one docker EMQX 4.4.18 exposing port 8084

Using this testing environment, I have a small app (in golang, using paho mqtt client) that creates two mqtt client, one connected to EMQX v5 using WS protocol, and another connected to EMQX v4 using WS protocol.

Then with each client I simply do subscribe, unsubscribe, subscribe.

My issue, and where I need some help/support, is in analysing the network packets using wireshark.

If I do a filter tcp.port == 8084 && websocket, I can properly see the exchangse between my client and EMQX 4. I can see the payload and I can visualize the sub/suback and unsub/unsuback messages.

But when I do a filter tcp.port == 8083 && websocket, I don't see any exchange between my client and EMQX 5. If I limit the filter to just tcp.port == 8083, then I can see all the TCP communication between my client and the broker, but I cannot interpret any frame.

What am I doing wrong? Is it a possible bug with MQTT?

My environment:

* win 10
* wireshark 4.0.6 + npcap (I listen on the loopback interface as everything is local to my laptop)
* client in golang (1.20) using paho golang client (v1.4.5)
* emqx v4 is 4.4.18, default conf (docker)
* emqx v5 is latest docker image, default conf

I have attached my wireshark capture as a zip file.

Here is a screenshot of the filter for EMQX v4 communications:
![image](https://github.com/emqx/emqx/assets/1410076/73d9e444-cabc-40fe-b494-57013e3f9fba)

Here is what it looks like for EMQX v5 communications (no websocket, I cannot understand the TCP communication):
![image](https://github.com/emqx/emqx/assets/1410076/0483dc6e-16b8-48a8-9089-68187493dee0)

Any help is apreciated!

[wireshark_NPF_LoopbackNSLC71.zip](https://github.com/emqx/emqx/files/11892736/wireshark_NPF_LoopbackNSLC71.zip)

**Comment by id** [comment]
@ldamet could you give us the exact command you use to run emqx 5.1?

**Reply by ldamet** [reply]
sure, here are my docker compose files for emqx 5 and emqx 4:

``` version: '3.8'

services:

  mqtt:
    image: emqx/emqx
    container_name: emqx
    environment:
      - EMQX_LOG__CONSOLE_HANDLER__LEVEL=debug
    restart: unless-stopped
    ports:
     - 1883:1883
     - 8083:8083
     - 18083:18083

networks:
  default:
    external: true
    name: local-dev
```
```
version: '3.8'

services:

  mqtt_old:
    image: emqx/emqx:4.4.18
    container_name: emqx_old
    environment:
      - EMQX_LOG__LEVEL=debug
      - EMQX_LOG__TO=both
    restart: unless-stopped
    ports:
     - 1884:1883
     - 8084:8083
     - 18084:18083

networks:
  default:
    external: true
    name: local-dev
```

Edit: in case you need it, the code used to connect/sub/unsub to the brokers:

```
package main

import (
        "fmt"
        "time"

        paho "github.com/eclipse/paho.mqtt.golang"
        "sdp-common-backend.git/log"
        "sdp-common-backend.git/mqtt"
        "sdp-common-backend.git/util"
)

func main() {
        fmt.Println("starting lda_mqtt...")
        log.SetLevel("debug")

        emqx5Client := mqtt.NewClient(&mqtt.Mq{
                URL: util.StringPtr("ws://localhost:8083/mqtt"),
                //URL: util.StringPtr("tcp://localhost:1883/mqtt"),
        }, "lda_mqtt_test_v5", nil)
        emqx4Client := mqtt.NewClient(&mqtt.Mq{
                //URL: util.StringPtr("tcp://localhost:1884/mqtt"),
                URL: util.StringPtr("ws://localhost:8084/mqtt"),
        }, "lda_mqtt_test_v4", nil)

        go func() {
                for {
                        time.Sleep(1000 * time.Millisecond)
                        //emqx5Client.SendMessageNR("lda/test2", `{"msg": "LDA EMQX V5"}`)
                }
        }()
        go func() {
                for {
                        time.Sleep(1000 * time.Millisecond)
                        //emqx4Client.SendMessageNR("lda/test2", `{"msg": "LDA EMQX V4"}`)
                }
        }()

        emqx5Client.Subscribe("lda/#", 1, emqx5SubHandler)
        emqx4Client.Subscribe("lda/#", 1, emqx4SubHandler)

        time.Sleep(10000 * time.Millisecond)

        fmt.Println("*************** RESUBSCRIBING *************")
        go func() {
                emqx5Client.Unsubscribe("lda/#")
                emqx5Client.Subscribe("lda/#", 1, emqx5SubHandler)
        }()

        go func() {
                emqx4Client.Unsubscribe("lda/#")
                emqx4Client.Subscribe("lda/#", 1, emqx4SubHandler)
        }()

        for {
                time.Sleep(1000 * time.Millisecond)
        }

        fmt.Println("exiting lda_mqtt!")
}

func emqx5SubHandler(client paho.Client, message paho.Message) {
        fmt.Printf("eqmx 5 handler received message: %s\n", string(message.Payload()))
}
func emqx4SubHandler(client paho.Client, message paho.Message) {
        fmt.Printf("eqmx 4 handler received message: %s\n", string(message.Payload()))
}
```

the custom mqtt package I use is just a wrapper around paho mqtt client:
```
opts := mqtt.NewClientOptions()
	opts.AddBroker(client.url)
	opts.SetClientID(backendID + "-producer-" + strconv.FormatInt(util.InstantEpochMilli(), 10))
	opts.SetAutoReconnect(true)
	opts.SetCleanSession(true)
	opts.SetConnectTimeout(10 * time.Second)
	if onConnHandler != nil {
		opts.SetOnConnectHandler(onConnHandler)
		opts.SetKeepAlive(5)
	}

	client.mqttClient = mqtt.NewClient(opts)
	token := client.mqttClient.Connect()
	for !token.WaitTimeout(3 * time.Second) {
	} 
```

**Comment by tkeksa** [comment, marked as answer]
Hi,
it works well in my Wireshark (tested 4.0.1 and last development version). It is probably some problem in your Wireshark settings.
When I set filter to tcp.port == 8083 I can see frame #4 as HTTP GET followed with switching to WS. But your screenshot shows that it tries to decode frame #4 as MQTT which is wrong. You have probably somewhere set that TCP port 8083 should be decoded as MQTT (either in MQTT protocol setting or in "Decode as ..." functionality). Try to disable it and it should work.

**Reply by ldamet** [reply]
Yes that's it! thanks a lot.

I had configured the MQTT protocol in wireshark to use 8083 as default port when I was testing some stuff some time ago. Switching it back to 1883 (tcp port), the capture correctly sees the switch to WS :)
